src/target/nrf91.c

```diff
@@ -81,25 +134,180 @@ static void nrf91_add_flash(target_s *target, uint32_t addr, size_t length, size
+			target->driver = "Nordic nRF91";
+	}
+
+	target->target_options |= TOPT_INHIBIT_NRST;
+	target_add_ram(target, 0x20000000, 256U * 1024U);
+	nrf91_add_flash(target, 0, 4096U * 256U, 4096U);
+
+	target->mass_erase = nrf91_mass_erase;
+
+
+	/* Persist AP access if uninitialized (only needed for devices with hardenend APPROTECT) */
+	/* TODO: can't add this, it breaks normal loading */
+//	if (ap->dp->target_revision > 2) {
+//		/* read UICR.APPROTECT */
+//		ap->dp->mem_read(ap, &buf[0], UICR_APPROTECT, sizeof(uint32_t));
+//		/* read UICR.SECUREAPPROTECT */
+//		ap->dp->mem_read(ap, &buf[1], UICR_SECUREAPPROTECT, sizeof(uint32_t));
+//
+//		target_mem_write32(target, NRF91_NVMC_CONFIG, NRF91_NVMC_CONFIG_WEN);
+//
+//		while (target_mem_read32(target, NRF91_NVMC_READY) == 0) {
+//			platform_delay(1);
+//			DEBUG_INFO("Waiting for NVMC to become ready\n");
+//		}
+//
+//		if (target_mem_read32(target, UICR_APPROTECT) == UICR_ERASED_VAL) {
+//			target_mem_write32(target, UICR_APPROTECT, UICR_APPROTECT_UNPROTECT_VAL);
+//		}
+//		if (target_mem_read32(target, UICR_SECUREAPPROTECT) == UICR_ERASED_VAL) {
+//			target_mem_write32(target, UICR_SECUREAPPROTECT, UICR_APPROTECT_UNPROTECT_VAL);
+//		}
+//
+//		target_mem_write32(target, NRF91_NVMC_CONFIG, NRF91_NVMC_CONFIG_REN);
+//	}
+
+	return true;
+}
+
+
+
+
+static bool nrf91_rescue_do_recover(target_s *t)
+{
+	adiv5_access_port_s *ap = (adiv5_access_port_s *)t->priv;
+
+	const bool hardened_approtect = ap->dp->target_revision > 2;
+
+	/* on some revisions, this needs to be repeated */
+	for (size_t i = 0; i < 3; ++i)
+	{
+		if (!nrf91_ctrl_ap_mass_erase(ap))
+			continue;
+		if (!hardened_approtect) {
+			/* pin reset is needed on older devices */
+			platform_nrst_set_val(true);
+			platform_delay(100);
+			platform_nrst_set_val(false);
+
+			/* repetition not needed and debug port inactive at this point */
+			return false;
+		}
+
+		//check if CSW DEVICEEN is set
+		struct adiv5_access_port ahb_ap = *ap;
+		ahb_ap.apsel = 0x0U;
+		uint32_t csw = ap->dp->ap_read(&ahb_ap, ADIV5_AP_CSW);
+		if (csw & ADIV5_AP_CSW_DEVICEEN) {
+			DEBUG_INFO("nRF91 Rescue succeeded.\n");
+			break;
+		}
+	}
+
+	return false;
+}
+
+bool nrf91_rescue_probe(adiv5_access_port_s *ap)
+{
+	target_s *t = target_new();
+	if (!t) {
 		return false;
 	}
+	adiv5_ap_ref(ap);
+	t->attach = (void *)nrf91_rescue_do_recover;
+	t->priv = ap;
+	t->priv_free = (void *)adiv5_ap_unref;
+	t->driver = "nRF91 Rescue (Attach, then scan again!)";
 
 	return true;
 }
+
+int nrf91_dp_prepare(adiv5_debug_port_s *const dp)
+{
+	adiv5_access_port_s ahb_ap = {
+		.dp = dp,
+		.apsel = 0x0U,
+	};
+	adiv5_access_port_s ctrl_ap = {
+		.dp = dp,
+		.apsel = 0x4U,
+	};
+	ahb_ap.idr = adiv5_ap_read(&ahb_ap, ADIV5_AP_IDR);
+	ahb_ap.csw = adiv5_ap_read(&ahb_ap, ADIV5_AP_CSW);
+	ctrl_ap.idr = adiv5_ap_read(&ctrl_ap, ADIV5_AP_IDR);
+
+	if (ahb_ap.idr != AHB_AP_IDR_EXPECTED) {
+		DEBUG_ERROR("nRF91: AHB-AP IDR is 0x%08x, expected 0x84770001\n", ahb_ap.idr);
```

Comment:
Please use `PRIx32` for the format specifier, not `x`, as idr is a `uint32_t` and so this doesn't work on Xtensa or several desktop platforms.

eg: `DEBUG_ERROR("nRF91: AHB-AP IDR is 0x%08" PRIx32 ", expected 0x84770001\n", ahb_ap.idr);`

It would also be better for Flash space if the expected value were formatted in the same way.